File: estructurales/proxy.java
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

interface ContenidoStreaming {
    void reproducir();
    String obtenerDetalles();
}

class ContenidoReal implements ContenidoStreaming {
    private String titulo;
    private int duracion;  // en minutos
    private String calidad;  // HD, 4K, etc.
    
    public ContenidoReal(String titulo, int duracion, String calidad) {
        this.titulo = titulo;
        this.duracion = duracion;
        this.calidad = calidad;
        System.out.println("Cargando contenido: " + titulo);
    }
    
    @Override
    public void reproducir() {
        System.out.println("Reproduciendo '" + titulo + "' en calidad " + calidad);
    }
    
    @Override
    public String obtenerDetalles() {
        return "Título: " + titulo + "\nDuración: " + duracion + " min\nCalidad: " + calidad;
    }
}

class ProxyContenido implements ContenidoStreaming {
    private String titulo;
    private int duracion;
    private String calidad;
    private String tipoSuscripcion;
    private ContenidoReal contenidoReal;
    private int contadorVisitas;
    
    public ProxyContenido(String titulo, int duracion, String calidad, String tipoSuscripcion) {
        this.titulo = titulo;
        this.duracion = duracion;
        this.calidad = calidad;
        this.tipoSuscripcion = tipoSuscripcion;
        this.contadorVisitas = 0;
    }
    
    private boolean verificarAcceso() {
        if (tipoSuscripcion.equals("free") && calidad.equals("4K")) {
            return false;
        } else if (tipoSuscripcion.equals("free") && contadorVisitas >= 3) {
            return false;
        } else if (tipoSuscripcion.equals("básico") && calidad.equals("4K")) {
            return false;
        } else {
            return true;
        }
    }
    
    @Override
    public void reproducir() {
        if (verificarAcceso()) {
            if (contenidoReal == null) {
                contenidoReal = new ContenidoReal(titulo, duracion, calidad);
            }
            
            LocalDateTime ahora = LocalDateTime.now();
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm:ss");
            String horaActual = ahora.format(formatter);
            System.out.println("[" + horaActual + "] Acceso registrado para análisis de uso");
            
            contadorVisitas++;
            contenidoReal.reproducir();
        } else {
            String calidadPermitida = tipoSuscripcion.equals("básico") ? "HD" : "SD";
            System.out.println("Acceso denegado a '" + titulo + "' en calidad " + calidad);
            System.out.println("Actualiza tu suscripción o elige calidad " + calidadPermitida);
        }
    }
    
    @Override
    public String obtenerDetalles() {
        System.out.println("Consulta de detalles por usuario con plan " + tipoSuscripcion);
        
        if (contenidoReal == null) {
            contenidoReal = new ContenidoReal(titulo, duracion, calidad);
        }
        
        return contenidoReal.obtenerDetalles();
    }
}

public class ProxyDemo {
    public static void main(String[] args) {
        System.out.println("=== USUARIO CON PLAN FREE ===");
        ProxyContenido pelicula1 = new ProxyContenido("El Código Enigma", 114, "4K", "free");
        pelicula1.reproducir();
        System.out.println("\nDetalles:");
        System.out.println(pelicula1.obtenerDetalles());
        
        System.out.println("\n=== USUARIO CON PLAN PREMIUM ===");
        ProxyContenido pelicula2 = new ProxyContenido("El Código Enigma", 114, "4K", "premium");
        pelicula2.reproducir();
        
        System.out.println("\n=== LÍMITE DE REPRODUCCIONES (PLAN FREE) ===");
        ProxyContenido documental = new ProxyContenido("Vida Salvaje", 45, "HD", "free");
        for (int i = 0; i < 4; i++) {
            System.out.println("\nIntento " + (i+1) + ":");
            documental.reproducir();
        }
    }
}

File: estructurales/proxy.py
from abc import ABC, abstractmethod
from datetime import datetime

class ContenidoStreaming(ABC):
    @abstractmethod
    def reproducir(self):
        pass
    
    @abstractmethod
    def obtener_detalles(self):
        pass

class ContenidoReal(ContenidoStreaming):
    def __init__(self, titulo, duracion, calidad):
        self.titulo = titulo
        self.duracion = duracion  # en minutos
        self.calidad = calidad    # HD, 4K, etc.
        print(f"Cargando contenido: {titulo}")
        
    def reproducir(self):
        print(f"Reproduciendo '{self.titulo}' en calidad {self.calidad}")
        
    def obtener_detalles(self):
        return f"Título: {self.titulo}\nDuración: {self.duracion} min\nCalidad: {self.calidad}"

class ProxyContenido(ContenidoStreaming):
    def __init__(self, titulo, duracion, calidad, tipo_suscripcion):
        self.titulo = titulo
        self.duracion = duracion
        self.calidad = calidad
        self.tipo_suscripcion = tipo_suscripcion
        self.contenido_real = None
        self.contador_visitas = 0
        
    def _verificar_acceso(self):
        # Diferentes niveles de suscripción tienen diferentes permisos
        if self.tipo_suscripcion == "free" and self.calidad == "4K":
            return False
        elif self.tipo_suscripcion == "free" and self.contador_visitas >= 3:
            return False
        elif self.tipo_suscripcion == "básico" and self.calidad == "4K":
            return False
        else:
            return True
            
    def reproducir(self):
        if self._verificar_acceso():
            if not self.contenido_real:
                self.contenido_real = ContenidoReal(self.titulo, self.duracion, self.calidad)
            
            # Registrar hora de visualización para análisis
            hora_actual = datetime.now().strftime("%H:%M:%S")
            print(f"[{hora_actual}] Acceso registrado para análisis de uso")
            
            self.contador_visitas += 1
            self.contenido_real.reproducir()
        else:
            calidad_permitida = "HD" if self.tipo_suscripcion == "básico" else "SD"
            print(f"Acceso denegado a '{self.titulo}' en calidad {self.calidad}")
            print(f"Actualiza tu suscripción o elige calidad {calidad_permitida}")
            
    def obtener_detalles(self):
        # Todos pueden ver los detalles, pero se registra quién lo consulta
        print(f"Consulta de detalles por usuario con plan {self.tipo_suscripcion}")
        
        if not self.contenido_real:
            self.contenido_real = ContenidoReal(self.titulo, self.duracion, self.calidad)
            
        return self.contenido_real.obtener_detalles()

# Ejemplo de uso
print("=== USUARIO CON PLAN FREE ===")
pelicula1 = ProxyContenido("El Código Enigma", 114, "4K", "free")
pelicula1.reproducir()
print("\nDetalles:")
print(pelicula1.obtener_detalles())

print("\n=== USUARIO CON PLAN PREMIUM ===")
pelicula2 = ProxyContenido("El Código Enigma", 114, "4K", "premium")
pelicula2.reproducir()

print("\n=== LÍMITE DE REPRODUCCIONES (PLAN FREE) ===")
documental = ProxyContenido("Vida Salvaje", 45, "HD", "free")
for i in range(4):
    print(f"\nIntento {i+1}:")
    documental.reproducir()
